File: articles/calculus-for-systems-modeling/articles/derivatives-and-rates-of-change/fortran/rate_diagnostics.f90
program rate_diagnostics
  implicit none
  real(8), dimension(5) :: hs
  real(8) :: x,h,exact,est
  integer :: i
  hs = (/1.0d0,0.5d0,0.25d0,0.125d0,0.0625d0/)
  x = 5.0d0
  exact = exact_derivative(x)
  print '(A)', 'method x0 h estimate exact absolute_error elasticity'
  do i=1,size(hs)
    h=hs(i)
    est=average_rate(x,x+h); call emit('average_rate_right',x,h,est,exact)
    est=forward_difference(x,h); call emit('forward_difference',x,h,est,exact)
    est=backward_difference(x,h); call emit('backward_difference',x,h,est,exact)
    est=central_difference(x,h); call emit('central_difference',x,h,est,exact)
  end do
contains
  real(8) function system_response(x)
    real(8), intent(in) :: x
    system_response=exp(0.2d0*x)
  end function
  real(8) function exact_derivative(x)
    real(8), intent(in) :: x
    exact_derivative=0.2d0*exp(0.2d0*x)
  end function
  real(8) function average_rate(a,b)
    real(8), intent(in) :: a,b
    average_rate=(system_response(b)-system_response(a))/(b-a)
  end function
  real(8) function forward_difference(x,h)
    real(8), intent(in) :: x,h
    forward_difference=(system_response(x+h)-system_response(x))/h
  end function
  real(8) function backward_difference(x,h)
    real(8), intent(in) :: x,h
    backward_difference=(system_response(x)-system_response(x-h))/h
  end function
  real(8) function central_difference(x,h)
    real(8), intent(in) :: x,h
    central_difference=(system_response(x+h)-system_response(x-h))/(2.0d0*h)
  end function
  real(8) function elasticity(d,x)
    real(8), intent(in) :: d,x
    elasticity=(x/system_response(x))*d
  end function
  subroutine emit(name,x,h,est,exact)
    character(len=*), intent(in) :: name
    real(8), intent(in) :: x,h,est,exact
    print '(A,1X,F8.4,1X,F8.4,1X,F14.8,1X,F14.8,1X,F14.8,1X,F14.8)', trim(name),x,h,est,exact,abs(est-exact),elasticity(est,x)
  end subroutine
end program

File: articles/calculus-for-systems-modeling/articles/derivatives-and-rates-of-change/r/rate_diagnostics.R
system_response <- function(x) exp(0.2 * x)
exact_derivative <- function(x) 0.2 * exp(0.2 * x)
average_rate <- function(a, b) (system_response(b) - system_response(a)) / (b - a)
forward_difference <- function(x, h) (system_response(x + h) - system_response(x)) / h
backward_difference <- function(x, h) (system_response(x) - system_response(x - h)) / h
central_difference <- function(x, h) (system_response(x + h) - system_response(x - h)) / (2 * h)
elasticity <- function(d, x) (x / system_response(x)) * d

x0 <- 5.0
h_values <- c(1, 0.5, 0.25, 0.125, 0.0625)
exact <- exact_derivative(x0)

results <- data.frame(
  x0 = x0,
  h = h_values,
  average_rate_right = sapply(h_values, function(h) average_rate(x0, x0 + h)),
  forward_difference = sapply(h_values, function(h) forward_difference(x0, h)),
  backward_difference = sapply(h_values, function(h) backward_difference(x0, h)),
  central_difference = sapply(h_values, function(h) central_difference(x0, h)),
  exact_derivative = exact
)
results$central_absolute_error <- abs(results$central_difference - results$exact_derivative)
results$elasticity_estimate <- sapply(results$central_difference, function(d) elasticity(d, x0))
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_rate_diagnostics.csv", row.names = FALSE)
print(results)
